[METRO-124] Create a new LanguageAdministrator plugin

Vinodsathyaseelan opened the pull request:
[METRO-124] Create a new LanguageAdministrator plugin

- Reason for change :  Enhancement
    Currently languages need to be set per application, which also does not work when the application is not active. 
	We should extend the Location sync to provide the current active langue/locale (by default determined by the location) and 
    the possibility to override this from the interface (e.g. when the middle ware allows the user to select a different language).  
			   
    Also a notification of this change/event should be available so JS can respsond to this event.
	 
- Solution : 
	Instead of extending LocationSync, we have agreed to introduce a new plugin named LanguageAdministrator
    to take care of the language related requirements.

pwielders commented on LanguageAdministrator/LanguageAdministrator.cpp:
```diff
@@ -0,0 +1,186 @@
+        }
+    }
+
+    void LanguageAdministrator::StateChange(PluginHost::IShell* plugin, const string& callsign)
```
Sorry, you where not aware of this, but yesterday a breaking change for this was added. Right now you need to split this method in 2 methods, one called void Activated (const string& callsign, PluginHost::IShell* plugin) and the other void Deactivated (const string& callsign, PluginHost::IShell* plugin).
The reason for this change is to avoid chatiness on getting the plugin->State() which might be over process boundaries and to make the reported states by Thunder less, only Activate and Deactivated are rported now..

pwielders commented on LanguageAdministrator/LanguageAdministratorImpl.cpp:
```diff
@@ -0,0 +1,102 @@
+    {
+        ASSERT(notification);
+        _adminLock.Lock();
+        //TODO: Store the language code to file.
```
old comment?

Vinodsathyaseelan replied:
Yes, old comment.  Will remove this.

pwielders commented on LanguageAdministrator/LanguageAdministratorImpl.cpp:
```diff
@@ -0,0 +1,102 @@
+        _adminLock.Unlock();
+    }
+
+    /*bool LanguageAdministratorImpl::IsvalidLanguageTag(const string& language) {
```
Can we remove dead code? The SCM will keep track of it, if you need it later on.

pwielders commented on LanguageAdministrator/LanguageAdministratorImpl.h:
```diff
@@ -0,0 +1,60 @@
+        void Register(Exchange::ILanguageTag::INotification* observer) override;
+        void Unregister(const Exchange::ILanguageTag::INotification* observer) override;
+
+        virtual uint32_t Language(string& language /* @out */) const override;
```
According to the coding rules and guidelines, you can only have virtual or override on a mehod, not both. Also, here you use string, which is good but in the interface specification you used std:string, which I reuested to change to string as well, than it is in line with this here..

Vinodsathyaseelan replied:
Forget to remove 'virtual' after copy paste from Interface definition. I will use sting instead of std::string.

pwielders commented on LanguageAdministrator/LanguageAdministratorImpl.h:
```diff
@@ -0,0 +1,60 @@
+        virtual uint32_t Language(const string& language) override;
+
+    private:
+        std::string _language;
```
Do not use the std on string, leave it up to Thunder to select the right flavour of string unless it *must* be a char or wchar representations...

pwielders reviewed (changes requested):
Looks good a few minors and a recent change which requires a small adaption of the code. But looks good !

Vinodsathyaseelan commented:
Implemented Activated() & Deactivated() functions and removed StateChange() in LanguageAdministrator plugin.
Addressed other review comments from Pierre Wielders.
